core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,56 +125,78 @@ public class JsPeripheral internal constructor(
-        _state.value = State.Connected
     }
 
-    private fun closeConnection() {
-        observationListeners.clear()
-        disconnectGatt()
-        unregisterDisconnectedListener()
-        _state.value = State.Disconnected()
+    private val disconnectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
+        connectJob.join()
+        closeConnection()
     }
 
     public override suspend fun connect() {
-        val job = connectJob ?: connectAsync().also { connectJob = it }
-        job.await()
+        connectJob.launch().join()
     }
 
     public override suspend fun disconnect() {
-        job.cancelAndJoinChildren()
-        connectJob = null
-        disconnectGatt()
+        connectJob.cancelAndJoin()
+        disconnectJob.launch().join()
+    }
+
+    private suspend fun openConnection() {
+        // Guard if already connected do nothing
+        if (_state.value is State.Connected) return
+
+        logger.info { message = "Connecting" }
+        // Wait until fully disconnected before proceeding
+        _state.first { it is State.Disconnected }
+
+        _state.value = State.Connecting.Bluetooth
+        registerDisconnectedListener()
+        gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
+
+        _state.value = State.Connecting.Services
+        discoverServices()
+        onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
+
+        _state.value = State.Connecting.Observes
+        logger.verbose { message = "Configuring characteristic observations" }
+        observers.onConnected()
+
+        logger.info { message = "Connected" }
+        _state.value = State.Connected
     }
 
-    private fun disconnectGatt() {
-        _state.value = State.Disconnecting
-        bluetoothDevice.gatt?.disconnect()
-        // Avoid trampling existing `Disconnected` state (and its properties) by only updating if not already `Disconnected`.
-        _state.update { previous -> previous as? State.Disconnected ?: State.Disconnected() }
+    private suspend fun closeConnection() {
+        // Guard if already disconnecting/disconnected do nothing
+        if (_state.value is State.Disconnecting || _state.value is State.Disconnected) return
+
+        try {
+            _state.value = State.Disconnecting
+            unregisterDisconnectedListener()
+            stopAllObservations()
+        } finally {
+            // IMPORTANT: update state FIRST as the gatt.disconnect may throw
+            _state.value = State.Disconnected()
+            bluetoothDevice.gatt?.disconnect()
+        }
+    }
+
+    private fun finalCleanup() {
+        try {
+            unregisterDisconnectedListener()
+            clearObservationsWithoutStopping()
+        } finally {
+            // IMPORTANT: update state FIRST as the gatt.disconnect may throw
+            _state.value = State.Disconnected()
+            bluetoothDevice.gatt?.disconnect()
```

Comment:
I'm worried that updating state first could cause other operations to continue (e.g. `_state.first { it is Disconnected }`) before the disconnect is kicked off?

Perhaps it would be better to nest another `try`-`finally`?

```suggestion
            try {
                bluetoothDevice.gatt?.disconnect()
            } finally {
                _state.value = State.Disconnected()
            }
```

Reply:
Yes you are right. Ugly, but probably the best approach here.